service/history/replication/executable_backfill_history_events_task.go

```diff
@@ -0,0 +1,322 @@
+		// baseExecutionInfo *workflowpb.BaseExecutionInfo
+
+		taskAttr *replicationspb.BackfillHistoryTaskAttributes
+
+		markPoisonPillAttempts int
+	}
+)
+
+var _ ctasks.Task = (*ExecutableBackfillHistoryEventsTask)(nil)
+var _ TrackableExecutableTask = (*ExecutableBackfillHistoryEventsTask)(nil)
+
+func NewExecutableBackfillHistoryEventsTask(
+	processToolBox ProcessToolBox,
+	taskID int64,
+	taskCreationTime time.Time,
+	task *replicationspb.BackfillHistoryTaskAttributes,
+	sourceClusterName string,
+	priority enumsspb.TaskPriority,
+) *ExecutableBackfillHistoryEventsTask {
+	return &ExecutableBackfillHistoryEventsTask{
+		ProcessToolBox: processToolBox,
+
+		WorkflowKey: definition.NewWorkflowKey(task.NamespaceId, task.WorkflowId, task.RunId),
+		ExecutableTask: NewExecutableTask(
+			processToolBox,
+			taskID,
+			metrics.BackfillHistoryEventsTaskScope,
+			taskCreationTime,
+			time.Now().UTC(),
+			sourceClusterName,
+			priority,
+		),
+		taskAttr:               task,
+		markPoisonPillAttempts: 0,
+	}
+}
+
+func (e *ExecutableBackfillHistoryEventsTask) QueueID() interface{} {
+	return e.WorkflowKey
+}
+
+func (e *ExecutableBackfillHistoryEventsTask) Execute() error {
+	if e.TerminalState() {
+		return nil
+	}
+
+	namespaceName, apply, nsError := e.GetNamespaceInfo(headers.SetCallerInfo(
+		context.Background(),
+		headers.SystemPreemptableCallerInfo,
+	), e.NamespaceID)
+	if nsError != nil {
+		return nsError
+	} else if !apply {
+		e.Logger.Warn("Skipping the replication task",
+			tag.WorkflowNamespaceID(e.NamespaceID),
+			tag.WorkflowID(e.WorkflowID),
+			tag.WorkflowRunID(e.RunID),
+			tag.TaskID(e.ExecutableTask.TaskID()),
+		)
+		metrics.ReplicationTasksSkipped.With(e.MetricsHandler).Record(
+			1,
+			metrics.OperationTag(metrics.BackfillHistoryEventsTaskScope),
+			metrics.NamespaceTag(namespaceName),
+		)
+		return nil
+	}
+
+	ctx, cancel := newTaskContext(namespaceName, e.Config.ReplicationTaskApplyTimeout())
+	defer cancel()
+
+	shardContext, err := e.ShardController.GetShardByNamespaceWorkflow(
+		namespace.ID(e.NamespaceID),
+		e.WorkflowID,
+	)
+	if err != nil {
+		return err
+	}
+
+	engine, err := shardContext.GetEngine(ctx)
+	if err != nil {
+		return err
+	}
+
+	events, newRunEvents, err := e.getDeserializedEvents()
+	if err != nil {
+		return err
+	}
+
+	return engine.BackfillHistoryEvents(ctx, &shard.BackfillHistoryEventsRequest{
+		WorkflowKey:         e.WorkflowKey,
+		SourceClusterName:   e.SourceClusterName(),
+		VersionedHistory:    e.taskAttr.VersionedTransition,
+		BaseExecutionInfo:   e.taskAttr.BaseExecutionInfo,
+		VersionHistoryItems: e.taskAttr.VersionHistoryItems,
+		Events:              events,
+		NewEvents:           newRunEvents,
+		NewRunID:            e.taskAttr.NewRunId,
+	})
+
+}
+
+func (e *ExecutableBackfillHistoryEventsTask) HandleErr(err error) error {
+	switch retryErr := err.(type) {
+	case nil, *serviceerror.NotFound:
+		return nil
+	case *serviceerrors.SyncState:
+		// TODO: call SyncState.
```

Comment:
I guess we can add a method on the executable similar to Resend for the SyncState operation. It will be reused by other tasks as well.

Reply:
added a new SyncState() method (to be implemented once we have the API ready)